src/ComputeSharp.D2D1.SourceGenerators/Diagnostics/Analyzers/InvalidD2DInputArgumentAnalyzer.cs

```diff
@@ -0,0 +1,166 @@
+            {
+                return;
+            }
+
+            // We want to register a callback for each method invocation, as all 'D2D' intrinsics would map to one
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation operation = (IInvocationOperation)context.Operation;
+
+                // Cheap initial filter: we only care about static methods from the 'D2D' type
+                if (operation.TargetMethod is not { IsStatic: true, ContainingType.Name: "D2D" } targetMethodSymbol)
+                {
+                    return;
+                }
+
+                // Second cheap inital filter: we only care about invocations with a constant 'int' argument in first position.
+                // While we're validating this, let's also get the 'index' parameter, since we need to validate it anyway.
+                if (operation.Arguments is not [{ Value.ConstantValue: { HasValue: true, Value: int index } }, ..])
+                {
+                    return;
+                }
+
+                // Validate that the target method is one of the ones we care about, and get the target input type
+                if (!methodSymbols.TryGetValue(targetMethodSymbol, out D2D1PixelShaderInputType targetInputType))
+                {
+                    return;
+                }
+
+                // We have matched a target symbol, so let's try to get the parent shader
+                if (context.ContainingSymbol.FirstAncestorOrSelf<INamedTypeSymbol>() is not { TypeKind: TypeKind.Struct } typeSymbol)
+                {
+                    return;
+                }
+
+                // We found a containing type, make sure it has '[D2DInputCount]'
+                if (!typeSymbol.HasAttributeWithType(d2DInputCountAttributeSymbol))
+                {
+                    return;
+                }
+
+                // At this point we can assume the shader type is mostly valid: let's get the actual input counts and types
+                D2DPixelShaderDescriptorGenerator.InputTypes.GetInfo(
+                    typeSymbol,
+                    out int inputCount,
+                    out _,
+                    out _,
+                    out ImmutableArray<uint> inputTypes);
+
+                // First validation: the index must be in range
+                if ((uint)index >= (uint)inputCount)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        IndexOutOfRangeForD2DIntrinsic,
+                        operation.Syntax.GetLocation(),
+                        targetMethodSymbol.Name,
+                        index,
+                        typeSymbol,
+                        inputCount));
+                }
+                else if ((D2D1PixelShaderInputType)inputTypes[index] != targetInputType)
```

Comment:
Is this going to report an error if I use a simple input method on a complex input? Because that's definitely valid. It's only bad if you use a complex input method with a simple input.

Reply:
> "Because that's definitely valid."

Mmh where does the spec say it's valid? The only method would be [D2DGetInput](https://learn.microsoft.com/en-us/windows/win32/direct2d/d2dgetinput), but the docs say this:
> "Only available for simple inputs."

Doesn't seem to me like they're implying you should use this on complex inputs at all 🤔

Reply:
Hmm you're right! I'm pretty sure it works on complex inputs though. Should run this analyzer on PDN to make sure there aren't any false error reports.

Reply:
Please do and let me know if you have any hits. But then again even if you did have any hits, would you still want to rely on this if it's technically undefined behavior? 🤔

Reply:
I don't know, we'll just have to see what happens and where

Reply:
It may be instructive to open up `d2d1effecthelpers.hlsli` and see what it does for `D2DGetInput()` for complex inputs. If it's legal there, it should be legal in CS.D2D1.

Reply:
I've checked the header but there's a ton of weird conditions based on linking mode and whatnot, it's not really clear to me what it's doing. But also regardless of that, even if it _is_ currently using code that happens to be fine for both inputs, the fact the docs explicitly say it's not supported seems sus. I suppose I could try asking some D2D people next week perhaps 🤔